### PR TITLE
[vcpkg scripts] Crosscompiling fixes

### PR DESCRIPTION
**Pull request by Neumann-A**

### CODE REVIEW

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 1]
```diff
@@ -8,85 +7,45 @@ function(z_vcpkg_append_proglist var_to_append additional_binaries)
     foreach(prog IN LISTS proglist)
         if(VCPKG_DETECTED_CMAKE_${prog})
             if(meson_${prog})
-                string(APPEND "${var_to_append}" "${meson_${prog}} = '${VCPKG_DETECTED_CMAKE_${prog}}'\n")
+                string(TOUPPER "MESON_${meson_${prog}}" var_to_set)
+                set("${var_to_set}" "${meson_${prog}} = ['${VCPKG_DETECTED_CMAKE_${prog}}']" PARENT_SCOPE)
+            elseif(${prog} STREQUAL AR AND VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${config_type})
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:12]: probably need to do something similar with RC flags

**Comment by Neumann-A** [comment]
@m-kuhn any other additions necessary?

**Comment by m-kuhn** [comment]
Sending it through the build pipeline now

**Comment by m-kuhn** [comment]
Looks good, minimal additional changes are needed for iOS, either in here or as a follow up

https://github.com/Neumann-A/vcpkg/pull/9

**Comment by Neumann-A** [comment]
@m-kuhn does the `CMAKE_${lang}_OSX_DEPLOYMENT_TARGET_FLAG` change work for you?

**Comment by m-kuhn** [comment]
> @m-kuhn does the `CMAKE_${lang}_OSX_DEPLOYMENT_TARGET_FLAG` change work for you?

Yes, this looks good.

Sidenote: it took me an extra run to figure out that a change to `scripts/get_cmake_vars/CMakeLists.txt` does not trigger a rebuild, so I had to fake a modification into `scripts/cmake/vcpkg_configure_make.cmake` to force rebuild.

File: ports/vcpkg-cmake-get-vars/cmake_get_vars/CMakeLists.txt [hunk 2]
```diff
@@ -101,20 +102,21 @@ if(APPLE)
...
-        string(APPEND EXTRA_FLAGS " ${CMAKE_CXX_COMPILE_OPTIONS_TARGET} \"${CMAKE_CXX_COMPILER_TARGET}\"")
+foreach(lang IN_LIST VCPKG_LANGUAGES)
+    if(CMAKE_${lang}_COMPILER_TARGET)
+        if (CMAKE_${lang}_COMPILER_ID STREQUAL Clang)
+            string(PREPEND EXTRA_FLAGS "${CMAKE_${lang}_COMPILE_OPTIONS_TARGET}${CMAKE_${lang}_COMPILER_TARGET} ")
+        elseif(CMAKE_${lang}_COMPILE_OPTIONS_TARGET)
+            string(PREPEND EXTRA_FLAGS " ${CMAKE_${lang}_COMPILE_OPTIONS_TARGET} ${CMAKE_${lang}_COMPILER_TARGET}")
+        endif()
     endif()
 endif()
```
Comment (Neumann-A) [inline ports/vcpkg-cmake-get-vars/cmake_get_vars/CMakeLists.txt:118]:
```suggestion
```

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 3]
```diff
@@ -74,8 +74,8 @@ macro(z_vcpkg_determine_autotools_target_arch_mac out_var)
 endmacro()
 
 macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
-    string(REGEX MATCHALL "( |^)-D[^ ]+" CPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
-    string(REGEX MATCHALL "( |^)-D[^ ]+" CXXPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${flag_suffix}}")
+    string(REGEX MATCHALL "( |^)(-D|-isysroot|--sysroot=|-isystem|-m?[Aa][Rr][Cc][Hh]) ?[^ ]+" CPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
+    string(REGEX MATCHALL "( |^)(-D|-isysroot|--sysroot=|-isystem|-m?[Aa][Rr][Cc][Hh]) ?[^ ]+" CXXPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${flag_suffix}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_make.cmake:78]: @m-kuhn doesn't this also need `--target(=| )` ?

Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: I originally had it on ldflags also but it ran fine without in the recent tests. I cannot think of a reason for this to be present in cppflags but ultimately i only believe successful builds when it comes to autoconf based builds.

Let me test again

**Comment by Neumann-A** [comment]
@JackBoosY another baseline regression...

**Comment by JackBoosY** [comment]
> @JackBoosY another baseline regression...

Fixing in https://github.com/microsoft/vcpkg/pull/26783.

File: scripts/cmake/vcpkg_configure_make.cmake (hunk 3 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: >  I cannot think of a reason for this to be present in cppflags

cpp configure results might depend on it? I assume the correct regex is `...|--target=|-target|...` (similar to sysroot) if I remember the platform stuff in cmake correct.

Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: Finally got around to testing your branch, you are correct.
Also `LDFLAGS` will need it otherwise `libpq` fails with

```
ld: error: pg_config.o is incompatible with elf64-x86-64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 4]
```diff
@@ -108,10 +110,29 @@ macro(_vcpkg_adjust_flags flag_var)
...
         endif()
     endif()
+    set(target_param "")
+    list(GET VCPKG_LANGUAGES 0 lang)
+    if(CMAKE_${lang}_COMPILER_TARGET AND "${flag_var}" STREQUAL "CMAKE_${lang}_FLAGS")
+        if (CMAKE_${lang}_COMPILER_ID STREQUAL Clang)
+            string(PREPEND ${flag_var} "${CMAKE_${lang}_COMPILE_OPTIONS_TARGET}${CMAKE_${lang}_COMPILER_TARGET} ")
+            set(target_param "${CMAKE_${lang}_COMPILE_OPTIONS_TARGET}${CMAKE_${lang}_COMPILER_TARGET}")
+        elseif(CMAKE_${lang}_COMPILE_OPTIONS_TARGET)
+            string(PREPEND ${flag_var} " ${CMAKE_${lang}_COMPILE_OPTIONS_TARGET} ${CMAKE_${lang}_COMPILER_TARGET}")
```
Comment (m-kuhn) [inline scripts/get_cmake_vars/CMakeLists.txt:126]:
```suggestion
            string(PREPEND ${flag_var} "${CMAKE_${lang}_COMPILE_OPTIONS_TARGET} ${CMAKE_${lang}_COMPILER_TARGET} ")

```

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 5]
```diff
@@ -108,10 +110,30 @@ macro(_vcpkg_adjust_flags flag_var)
...
             if (CMAKE_OSX_DEPLOYMENT_TARGET)
-                string(APPEND ${flag_var} " -mmacosx-version-min=${CMAKE_OSX_DEPLOYMENT_TARGET}") 
+                list(GET VCPKG_LANGUAGES 0 lang)
+                string(APPEND ${flag_var} " ${CMAKE_${lang}_OSX_DEPLOYMENT_TARGET_FLAG}${CMAKE_OSX_DEPLOYMENT_TARGET}")
+                unset(lang)
             endif()
         endif()
     endif()
+    set(target_param "")
+    list(GET VCPKG_LANGUAGES 0 lang)
```
Comment (m-kuhn) [inline scripts/get_cmake_vars/CMakeLists.txt:120]: Why only the first one? This will prepend to `CMAKE_C_FLAGS` but not `CMAKE_CXX_FLAGS`.

Comment (Neumann-A) [reply scripts/get_cmake_vars/CMakeLists.txt:120]: this is looping over all vars

Comment (Neumann-A) [reply scripts/get_cmake_vars/CMakeLists.txt:120]: meaning the call to _vcpkg_adjust_flags is within a loop looping over all the vars and target_param should be the same for all languages.

Comment (m-kuhn) [reply scripts/get_cmake_vars/CMakeLists.txt:120]: if `flag_var` is `CMAKE_CXX_FLAGS` on line 121 `"${flag_var}" STREQUAL "CMAKE_${lang}_FLAGS"` will be false and line 123 is not executed and `--target` is not set.

Comment (Neumann-A) [reply scripts/get_cmake_vars/CMakeLists.txt:121]: Ok added a `flags_to_add_target` list and removed `target_param` since it isn't necessary any more.

File: ports/vcpkg-cmake-get-vars/cmake_get_vars/CMakeLists.txt [hunk 6]
```diff
@@ -101,20 +102,21 @@ if(APPLE)
...
+        unset(lang)
     endif()
 endif()
-if(CMAKE_CXX_COMPILER_TARGET)
-    if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-        string(APPEND EXTRA_FLAGS " \"${CMAKE_CXX_COMPILE_OPTIONS_TARGET}${CMAKE_CXX_COMPILER_TARGET}\"")
-    elseif(CMAKE_CXX_COMPILE_OPTIONS_TARGET)
-        string(APPEND EXTRA_FLAGS " ${CMAKE_CXX_COMPILE_OPTIONS_TARGET} \"${CMAKE_CXX_COMPILER_TARGET}\"")
+set(target_param "")
+foreach(lang IN_LIST VCPKG_LANGUAGES)
```
Comment (m-kuhn) [inline ports/vcpkg-cmake-get-vars/cmake_get_vars/CMakeLists.txt:111]:
```suggestion
list(GET VCPKG_LANGUAGES 0 lang)
```

This loop should also be replaced

1. it's only taking one `target_param`
2. `foreach` doesn't work with `IN_LIST` :see_no_evil:

Comment (m-kuhn) [reply ports/vcpkg-cmake-get-vars/cmake_get_vars/CMakeLists.txt:111]: https://github.com/Neumann-A/vcpkg/pull/12

**Review by m-kuhn — DISMISSED** [review]
Works well for android and ios now. Huge improvements 👍

**Review by LilyWangLL — DISMISSED** [review]

**Comment by LilyWangLL** [comment]
@Neumann-A Could you please resolve the conflicts?

**Review by JackBoosY — DISMISSED** [review]
I think we need further review by other team members.

**Review by LilyWangLL — DISMISSED** [review]

**Comment by Neumann-A** [comment]
Hmm don't know what happened to jemalloc on osx. Has been green in https://github.com/microsoft/vcpkg/pull/26617/commits/184abe3c41315e6a353b0016eb8ebce38c372b92 and only thing I did was merge with master. Looks like a concurrency issue?

**Comment by dg0yt** [comment]
> Hmm don't know what happened to jemalloc on osx. Has been green in [184abe3](https://github.com/microsoft/vcpkg/commit/184abe3c41315e6a353b0016eb8ebce38c372b92) and only thing I did was merge with master. Looks like a concurrency issue?

> ~~~
> install: mkdir /Users/vagrant/Data/packages/jemalloc_x64-osx/Users/vagrant/Data/installed: File exists

This is a typical symptom of concurrency issues during `make install` on osx. We need a `DISABLE_PARALLEL_INSTALL` and/or a reliable `mkdir` wrapper for osx.

**Comment by JackBoosY** [comment]
Should we wait for #27105 merge or just rerun the pipeline test?

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 7]
```diff
@@ -74,8 +74,8 @@ macro(z_vcpkg_determine_autotools_target_arch_mac out_var)
 endmacro()
 
 macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
-    string(REGEX MATCHALL "( |^)-D[^ ]+" CPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
-    string(REGEX MATCHALL "( |^)-D[^ ]+" CXXPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${flag_suffix}}")
+    string(REGEX MATCHALL "( |^)(-D|-isysroot|--sysroot=|-isystem|-m?[Aa][Rr][Cc][Hh]|--target=|-target) ?[^ ]+" CPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
+    string(REGEX MATCHALL "( |^)(-D|-isysroot|--sysroot=|-isystem|-m?[Aa][Rr][Cc][Hh]|--target=|-target) ?[^ ]+" CXXPPFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${flag_suffix}}")
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_configure_make.cmake:78]: @Neumann-A Do you remember why you remove `--target` from `CFLAGS` here? 
It is explicitly injected into the flags by `scripts/get_cmake_vars/CMakeLists.txt`.
Removing this flag breaks android toolchains. (CMake selects `clang` for compiler, so `--target=x86_64-none-linux-android21` etc. is needed to pick the right includes.) LDFLAGS still carries the option.

Comment (dg0yt) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: From building luajit (WIP):
~~~
ld: error: undefined symbol: _IO_putc
>>> referenced by luajit.c:150 (/home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/luajit/x64-android-dbg/src/luajit.c:150)
>>>               luajit.o:(print_jit_status)
>>> referenced by luajit.c:153 (/home/dg0yt/Projekte/vcpkg/vcpkg/buildtrees/luajit/x64-android-dbg/src/luajit.c:153)
>>>               luajit.o:(print_jit_status)
>>> referenced by stdio.h:81 (/usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../x86_64-linux-gnu/include/bits/stdio.h:81)
>>>               lib_base.o:(lj_cf_print) in archive libluajit.a
>>> referenced 1 more times
~~~
some time after:
~~~
/<android-ndk-r25c>/toolchains/llvm/prebuilt/linux-x86_64/bin/clang  -O2 -fomit-frame-pointer -Wall   -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -U_FORTIFY_SOURCE  -DLUA_ROOT=\"/home/dg0yt/Projekte/vcpkg/vcpkg/installed/x64-android/debug\" -DLUA_MULTILIB=\"lib\" -fno-stack-protector -DLUAJIT_UNWIND_EXTERNAL  -g -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -Wformat -Werror=format-security -fPIC -fno-limit-debug-info -c -o luajit.o luajit.c
~~~

Comment (dg0yt) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: Ignore. I found the right track now.
TLDR: There is CFLAGS and there is CPPFLAGS, and I should use both...

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_make.cmake:78]: Some people (meson) would argue all flags have to go into CC / CXX instead.
